### PR TITLE
[CXH-1572] - Add license management via workspace roles - Notion Connector

### PR DESCRIPTION
**Pull request by mateoHernandez123**
## Description
- [ ] Bug fix
- [x] New feature

Adds Notion workspace roles as a first-class C1 `Role` resource with grant / revoke provisioning, plus an in-process SCIM mock server so local dev and CI can run without an Enterprise Notion tenant. Closes [CXH-1572 — baton-notion: add license management](https://linear.app/ductone/issue/CXH-1572/baton-notion-add-license-management).

Notion's SCIM API does not expose a `/Licenses` endpoint — "license" in Notion product terminology means "paid seat" and is binary (member vs. guest), not tiered. The connector therefore models the **workspace role** itself as the license tier: each `Role` resource carries `TRAIT_LICENSE_PROFILE` alongside `TRAIT_ROLE` (Adobe-style hybrid, mirroring [baton-adobe #34](https://github.com/ConductorOne/baton-adobe/pull/34)), so every role doubles as a seat tier in C1 App Utilization.

### Sync:

- **Users** (`user`) — unchanged surface. Internally the SCIM extension's `role` value is stashed on the user profile during `List` so role grants can be emitted principal-side from `userBuilder.Grants` (one grant per user) without re-paginating users per role. Keeps sync at `O(users)` instead of `O(users × roles)`.
- **Groups** (`group`) — unchanged surface.
- **Roles** (`role`, NEW) — static set of four (`owner`, `membership_admin`, `member`, `restricted_member`) matching the enum on the [`urn:ietf:params:scim:schemas:extension:notion:2.0:User.role`](https://www.notion.com/help/provision-users-and-groups-with-scim) extension. Per-resource `Entitlements` (not `StaticEntitlements`) because each role's display name and description vary — matches the canonical `TRAIT_ROLE` pattern. Declares `TRAIT_LICENSE_PROFILE` with `WithLicenseEntitlementIDs` so seat-holders correlate to their grants for App Utilization. Seat counts (`WithLicenseSeats`) intentionally omitted — Notion's SCIM API does not expose a per-role purchased / consumed endpoint.

### Provisioning:

- **Create / Delete user accounts** — unchanged.
- **Grant / Revoke role assignment** — NEW. Both go through `PATCH /scim/v2/Users/{id}` on the Notion role extension. Idempotent via a single pre-flight `GET /Users/{id}`:
  - `GrantAlreadyExists` when the user already holds the target role (no PATCH issued).
  - `GrantAlreadyRevoked` on three short-circuits: (a) the user is gone (404), (b) the user no longer holds the role, (c) the role being revoked is `restricted_member` — Notion users must always carry a role, so revoke is modelled as a downgrade to `restricted_member` and the floor tier has nothing to downgrade to (account must be deprovisioned to remove the user from the workspace).

### Auth:

Unchanged. Existing `--scim-token` flag, no new credentials, no breaking changes for existing deployments.

### Architecture highlights:

- **Hybrid trait shape (`TRAIT_ROLE` + `TRAIT_LICENSE_PROFILE`).** Lets the same resource power both UAR / JIT flows and App Utilization without inventing a second `/Licenses` resource for an API that doesn't have one.
- **Principal-side grant emission** for the derived `role` type — `roleBuilder.Grants` returns `nil` intentionally; the actual grants come from `userBuilder.Grants`, which reads the stashed `workspace_role` profile field. Avoids the classic `O(N × M)` Grants scan.
- **Floor-tier `Revoke` short-circuit.** Revoking `restricted_member` is a no-op (warn-logged) with `GrantAlreadyRevoked` — there is no lower tier to downgrade to.
- **New `pkg/client/helpers.go`** centralizing API error classification (`IsNotFoundError` against the gRPC `NotFound` status that `uhttp.BaseHttpClient` maps from HTTP 404).
- **In-process SCIM mock under `test-server/`** (HTTP server + Postman collection + README), since Notion's SCIM API is Enterprise-only. The mock implements `GET/POST/PATCH/DELETE /Users`, `GET /Groups`, `ServiceProviderConfig`, and `ResourceTypes`, with stable seed IDs for CI assertions and both SCIM PATCH spellings (value-as-object and path) accepted on `PATCH /Users/{id}`.
- **Documentation refactored to the four-tier hierarchy** — `README.md` (developer), `docs/connector.mdx` (customer-facing, with `AUTO-GENERATED` markers for Generate Baton Metadata), `docs/docs-info.md` (internal engineering), `test-server/README.md` (mock).

### Useful links:

- [Linear CXH-1572 — baton-notion: add license management](https://linear.app/ductone/issue/CXH-1572/baton-notion-add-license-management)
- [Notion — Provision users & groups with SCIM](https://www.notion.com/help/provision-users-and-groups-with-scim)
- [RFC 7644 §3.5.2 — SCIM 2.0 Modifying with PATCH](https://datatracker.ietf.org/doc/html/rfc7644#section-3.5.2)
- [baton-adobe #34 — Adobe-style `TRAIT_ROLE` + `TRAIT_LICENSE_PROFILE` hybrid](https://github.com/ConductorOne/baton-adobe/pull/34)

### CODE REVIEW

File: pkg/connector/role.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+// Grants is intentionally empty — role grants are emitted from userBuilder.Grants.
+func (b *roleBuilder) Grants(_ context.Context, _ *v2.Resource, _ rs.SyncOpAttrs) ([]*v2.Grant, *rs.SyncOpResults, error) {
+	return nil, nil, nil
+}
+
+// Grant assigns a Notion workspace role to a user. Idempotent: if the user
+// already holds the target role, returns GrantAlreadyExists without issuing
+// a PATCH; a 409 from the PATCH is also treated as success.
```
[the remark this answers is not in the document]
Comment (mateoHernandez123) [reply pkg/connector/role.go]: Fixed in 32c0dc6 — removed the stale 409 claim from the doc comment. The code intentionally does not handle 409 on the PATCH: Notion's SCIM API only returns 409 (uniqueness) for duplicate userName on POST /Users, not for role updates, so the guard would be dead code. The pre-flight GET covers the idempotency path.

File: pkg/connector/role.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+func isSupportedRole(roleID string) bool {
+	for _, role := range supportedRoles {
+		if role.id == roleID {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (mateoHernandez123) [reply pkg/connector/role.go:275]: The provisioning edge cases (already-exists grant, floor-tier revoke, user-deleted-between-sync-and-revoke, unknown role skip) are covered by the end-to-end lifecycle run against the in-process SCIM mock (test-server/), which exercises the same branches through the real connector binary. We opted not to ship unit tests for these paths in this PR to keep the diff focused; happy to add them as a follow-up if reviewers feel strongly.

File: pkg/connector/role.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+		return nil, nil, fmt.Errorf("baton-notion: grant role %s to user %s (fetch current): %w", targetRole, userID, err)
+	}
+
+	if current.Role() == targetRole {
+		return nil, annotations.New(&v2.GrantAlreadyExists{}), nil
+	}
+
+	if _, err := b.client.PatchUserRole(ctx, userID, targetRole); err != nil {
+		return nil, nil, fmt.Errorf("baton-notion: grant role %s to user %s: %w", targetRole, userID, err)
+	}
```
Comment (laurenleach) [inline pkg/connector/role.go:210]: Since a user can only hold one role, we should use exclusion groups 
https://github.com/ConductorOne/baton-sdk/pull/765/changes#diff-000f935e03452547e61f12628d489f96d3da3805c3ff195622404987536ba0bbR58

and return the `GrantReplaced` annotation 
https://github.com/ConductorOne/baton-sdk/pull/765/changes#diff-b4a4de9aaed45993d9080d372e526e92c0d34bd95cc06436fb27891abedff5b5R20

**Review by laurenleach — CHANGES_REQUESTED** [review]

File: pkg/connector/role.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+func (b *roleBuilder) Entitlements(_ context.Context, resource *v2.Resource, _ rs.SyncOpAttrs) ([]*v2.Entitlement, *rs.SyncOpResults, error) {
+	description := fmt.Sprintf("Assigned the %s role in Notion", resource.DisplayName)
+	for _, role := range supportedRoles {
+		if role.id == resource.Id.Resource {
+			description = role.description
+			break
+		}
+	}
```
Comment (laurenleach) [inline pkg/connector/role.go]: I think we should use a map instead so we don't need to iterate an array for every role

File: pkg/connector/role.go (hunk 3 above)
Comment (FeliLucero1) [reply pkg/connector/role.go:210]: example https://github.com/ConductorOne/baton-brex/blob/5a3e8ac326edd446c64fb7894ce55235d39daa6b/pkg/connector/locations.go#L60

File: pkg/connector/role.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+			return annotations.New(&v2.GrantAlreadyRevoked{}), nil
+		}
+		return nil, fmt.Errorf("baton-notion: revoke role %s from user %s (fetch current): %w", targetRole, userID, err)
+	}
+
+	if current.Role() != targetRole {
+		return annotations.New(&v2.GrantAlreadyRevoked{}), nil
+	}
+
+	if targetRole == defaultRevokedRole {
```
Comment (FeliLucero1) [inline pkg/connector/role.go:254]: **Bug: revoking `restricted_member` returns `GrantAlreadyRevoked` while the user remains in the workspace.**

When `targetRole == "restricted_member"` (the floor tier), this branch logs a warning and returns `GrantAlreadyRevoked` with `nil` error. C1 records the grant as revoked, but the user is still a `restricted_member` in Notion with full workspace access — a silent state divergence.

`GrantAlreadyRevoked` means "the grant no longer exists", which is false here. The grant still exists and the user is still a workspace member.

Consider returning an explicit error instead:
```go
return nil, fmt.Errorf("baton-notion: cannot revoke restricted_member role — it is the minimum Notion workspace role; deprovision the account to remove the user from the workspace")
```
This surfaces the limitation clearly in the C1 UI rather than silently marking the entitlement as removed.

File: pkg/client/client.go [hunk 6]
```diff
@@ -22,6 +22,11 @@ type NotionClient struct {
...
 }
 
+// GetUsers paginates `GET /scim/v2/Users` using Notion's startIndex/count
+// convention (startIndex is 1-indexed, count caps at 100).
+//
+// Doc: https://www.notion.com/help/provision-users-and-groups-with-scim
+// (section "Users" → `GET /Users`).
 func (c *NotionClient) GetUsers(ctx context.Context, pageOps PaginationOptions) ([]User, string, error) {
 	var nextPage string
 	requestURL := fmt.Sprint(c.baseURL, "/Users")
```
Comment (FeliLucero1) [inline pkg/client/client.go]: **Bug: URL construction with `fmt.Sprint` produces double-slash paths when `baseURL` ends with `/`.**

`fmt.Sprint(c.baseURL, "/Users")` produces `.../scim/v2//Users` if the caller passes `--base-url http://localhost:8765/scim/v2/` (trailing slash), causing 404s on every call.

This is inconsistent with `PatchUserRole` (line 149) which correctly uses `url.JoinPath`. The same pattern recurs on lines 62, 91, 110, 131, and 180.

Fix all six occurrences:
```go
requestURL, err := url.JoinPath(c.baseURL, "Users")
if err != nil {
    return nil, "", fmt.Errorf("baton-notion: build URL: %w", err)
}
```

File: pkg/client/client.go [hunk 7]
```diff
@@ -125,8 +191,8 @@ func (c *NotionClient) doRequest(
 	ctx context.Context,
 	method string,
 	endpointUrl string,
-	res interface{},
-	body interface{},
+	res any,
+	body any,
```
Comment (FeliLucero1) [inline pkg/client/client.go:217]: **Bug (in function body below): GET and DELETE requests are sent without `Accept: application/json`.**

Further down in this function, `WithAcceptJSONHeader()` is only appended when `body != nil`, so all GET calls (list users, list groups, get user, get group) go out without an `Accept` header. If Notion's gateway returns `text/html` on an error page (e.g. a WAF-level 401), `uhttp.WithResponse` silently decodes it to a zero-value struct and returns `nil` error — producing an apparent empty sync with no visible failure.

Fix: always add `WithAcceptJSONHeader()` regardless of body:
```go
opts := []uhttp.RequestOption{
    uhttp.WithBearerToken(c.scimToken),
    uhttp.WithAcceptJSONHeader(),
}
if body != nil {
    opts = append(opts, uhttp.WithContentTypeJSONHeader(), uhttp.WithJSONBody(body))
}
```

File: pkg/connector/role.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+		logger.Warn(
```
Comment (FeliLucero1) [inline pkg/connector/role.go]: Debug?

File: pkg/connector/role.go (hunk 3 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go:210]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad) — used a cross-resource exclusion group with `GrantReplaced` on role change. Kept it **at-most-one** rather than mandatory/default: C1's default-replacement-on-revoke expects the default tier to already have a grant binding, which never holds here (a user has exactly one role at a time) and failed at runtime with `default replacement target binding has no connector grant id`.

File: pkg/connector/role.go (hunk 4 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad) — replaced the slice scans with a `rolesByID` map.

File: pkg/connector/role.go (hunk 5 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go:254]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad), with a tweak: handled as a `debug` log + `GrantAlreadyRevoked` (expected no-op) rather than an error. Revoking `restricted_member` is the floor tier (nothing to downgrade to), and erroring would surface a Warn/Error in the C1 UI on every such revoke. Happy to switch to an explicit error if you'd prefer to surface it.

File: pkg/client/client.go (hunk 6 above)
Comment (mateoHernandez123) [reply pkg/client/client.go]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad) — all six calls now use `url.JoinPath`.

File: pkg/client/client.go (hunk 7 above)
Comment (mateoHernandez123) [reply pkg/client/client.go:217]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad) — `doRequest` now always sends `WithAcceptJSONHeader()`.

File: pkg/connector/role.go (hunk 8 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go]: Done in [4ed8f50](https://github.com/ConductorOne/baton-notion/commit/4ed8f504ea30497d65ccd15cd8ffe0a5179d93ad) — changed to `debug`.

File: pkg/connector/role.go [hunk 9]
```diff
@@ -128,19 +146,23 @@ func (b *roleBuilder) List(_ context.Context, _ *v2.ResourceId, _ rs.SyncOpAttrs
...
 	}
 
 	options := []ent.EntitlementOption{
 		ent.WithGrantableTo(userResourceType),
 		ent.WithDisplayName(fmt.Sprintf("%s role %s", resource.DisplayName, assignedEntitlement)),
 		ent.WithDescription(description),
 	}
+
+	if ok {
+		options = append(options, ent.WithExclusionGroupOrder(workspaceRoleExclusionGroup, role.order))
```
Comment (laurenleach) [inline pkg/connector/role.go]: for `defaultRevokedRole` (`client.RoleRestrictedMember`), we should use `WithExclusionGroupDefault`

c1 will handle granting this entitlement if we revoke one of the other roles

File: pkg/connector/role.go [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+		return annotations.New(&v2.GrantAlreadyRevoked{}), nil
+	}
+
+	if targetRole == defaultRevokedRole {
+		l.Debug(
+			"baton-notion: cannot revoke restricted_member role — it is the floor tier; deprovision the account to remove the user from the workspace",
+			zap.String("user_id", userID),
+			zap.String("role_id", targetRole),
+		)
+		return annotations.New(&v2.GrantAlreadyRevoked{}), nil
```
Comment (FeliLucero1) [inline pkg/connector/role.go]: I think we should return an error here IMO

File: pkg/connector/role.go (hunk 9 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go]: Done in [f8ffbe4](https://github.com/ConductorOne/baton-notion/commit/f8ffbe46fd560675eda3b0416e91c7e1a54021af) — `restricted_member` now uses `WithExclusionGroupDefault`. Re-tested in the sandbox with a full sync first, and the revoke works end-to-end (no more `default replacement target binding` error). Thanks!

File: pkg/connector/role.go (hunk 10 above)
Comment (mateoHernandez123) [reply pkg/connector/role.go]: Done in [f8ffbe4](https://github.com/ConductorOne/baton-notion/commit/f8ffbe46fd560675eda3b0416e91c7e1a54021af) — revoking `restricted_member` now returns an explicit error instead of `GrantAlreadyRevoked` (it is the minimum role with nothing to downgrade to; the account must be deprovisioned).

**Review by laurenleach — APPROVED** [review]

**Review by FeliLucero1 — APPROVED** [review]
